# Pandas: Merge and groupby


In [1]:
import pandas as pd


In [2]:
df = pd.read_csv("data/card/aro_index.tsv", sep="\t")

In [3]:
df

,ARO Accession,CVTERM ID,Model Sequence ID,Model ID,Model Name,ARO Name,Protein Accession,DNA Accession,AMR Gene Family,Drug Class,Resistance Mechanism,CARD Short Name
0,ARO:3005099,43314,6143,3831,23S rRNA (adenine(2058)-N(6))-methyltransferas...,23S rRNA (adenine(2058)-N(6))-methyltransferas...,AAB60941.1,AF002716.1,Erm 23S ribosomal RNA methyltransferase,lincosamide antibiotic;macrolide antibiotic;st...,antibiotic target alteration,Spyo_ErmA_MLSb
1,ARO:3002523,38923,8144,1781,AAC(2')-Ia,AAC(2')-Ia,AAA03550.1,L06156.2,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ia
2,ARO:3002524,38924,85,746,AAC(2')-Ib,AAC(2')-Ib,AAC44793.1,U41471.1,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ib
3,ARO:3002525,38925,4719,1246,AAC(2')-Ic,AAC(2')-Ic,CCP42991.1,AL123456.3,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ic
4,ARO:3002526,38926,228,1415,AAC(2')-Id,AAC(2')-Id,AAB41701.1,U72743.1,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Id
...,...,...,...,...,...,...,...,...,...,...,...,...
6440,ARO:3003064,39498,536,350,ykkD,ykkD,CAB13167.1,AL009126.1,small multidrug resistance (SMR) antibiotic ef...,aminoglycoside antibiotic;phenicol antibiotic;...,antibiotic efflux,ykkD
6441,ARO:3009096,47888,11341,8514,YOC-1,YOC-1,QIU91361.1,CP050811.1,YOC beta-lactamase,cephalosporin,antibiotic inactivation,YOC-1
6442,ARO:3003952,40722,5261,2424,YojI,YojI,AAC75271.1,U00096.3,ATP-binding cassette (ABC) antibiotic efflux pump,peptide antibiotic,antibiotic efflux,YojI
6443,ARO:3005035,43233,6072,3779,YRC-1,YRC-1,ABA70720.1,DQ185144.1,YRC beta-lactamase,cephalosporin;penicillin beta-lactam,antibiotic inactivation,YRC-1


In [4]:
df.head()


,ARO Accession,CVTERM ID,Model Sequence ID,Model ID,Model Name,ARO Name,Protein Accession,DNA Accession,AMR Gene Family,Drug Class,Resistance Mechanism,CARD Short Name
0,ARO:3005099,43314,6143,3831,23S rRNA (adenine(2058)-N(6))-methyltransferas...,23S rRNA (adenine(2058)-N(6))-methyltransferas...,AAB60941.1,AF002716.1,Erm 23S ribosomal RNA methyltransferase,lincosamide antibiotic;macrolide antibiotic;st...,antibiotic target alteration,Spyo_ErmA_MLSb
1,ARO:3002523,38923,8144,1781,AAC(2')-Ia,AAC(2')-Ia,AAA03550.1,L06156.2,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ia
2,ARO:3002524,38924,85,746,AAC(2')-Ib,AAC(2')-Ib,AAC44793.1,U41471.1,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ib
3,ARO:3002525,38925,4719,1246,AAC(2')-Ic,AAC(2')-Ic,CCP42991.1,AL123456.3,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ic
4,ARO:3002526,38926,228,1415,AAC(2')-Id,AAC(2')-Id,AAB41701.1,U72743.1,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Id


In [6]:
df.columns


Index(['ARO Accession', 'CVTERM ID', 'Model Sequence ID', 'Model ID',
       'Model Name', 'ARO Name', 'Protein Accession', 'DNA Accession',
       'AMR Gene Family', 'Drug Class', 'Resistance Mechanism',
       'CARD Short Name'],
      dtype='str')

In [7]:
df.shape


(6445, 12)

In [14]:
counts = df.groupby("Drug Class").size().sort_values(ascending = False)
counts.head(15)

Drug Class
cephalosporin                                                 943
carbapenem;penicillin beta-lactam                             821
carbapenem;cephalosporin;monobactam                           638
penicillin beta-lactam                                        583
carbapenem;cephalosporin;penicillin beta-lactam               579
cephalosporin;penicillin beta-lactam                          461
cephalosporin;monobactam;penicillin beta-lactam               312
carbapenem;cephalosporin;monobactam;penicillin beta-lactam    287
aminoglycoside antibiotic                                     200
carbapenem                                                    184
peptide antibiotic                                            168
fluoroquinolone antibiotic                                    147
glycopeptide antibiotic                                        86
tetracycline antibiotic                                        82
cephalosporin;monobactam                                       74

In [16]:
# look at another CARD file
import os

os.listdir("data/card")

['protein_fasta_protein_variant_model.fasta',
 'nucleotide_fasta_protein_variant_model.fasta',
 'card-data.tar.bz2',
 'nucleotide_fasta_protein_overexpression_model.fasta',
 'Untitled.ipynb',
 'snps.txt',
 'protein_fasta_protein_knockout_model.fasta',
 'aro_categories.tsv',
 'nucleotide_fasta_protein_knockout_model.fasta',
 'card.json',
 'shortname_pathogens.tsv',
 'CARD-Download-README.txt',
 'protein_fasta_protein_overexpression_model.fasta',
 '.ipynb_checkpoints',
 'PMID.tsv',
 'nucleotide_fasta_rRNA_gene_variant_model.fasta',
 'aro_categories_index.tsv',
 'aro_index.tsv',
 'nucleotide_fasta_protein_homolog_model.fasta',
 'protein_fasta_protein_homolog_model.fasta',
 'shortname_antibiotics.tsv']

In [17]:
aro = pd.read_csv("data/card/aro_index.tsv", sep="\t")
categories = pd.read_csv("data/card/aro_categories.tsv", sep="\t")

In [18]:
aro

,ARO Accession,CVTERM ID,Model Sequence ID,Model ID,Model Name,ARO Name,Protein Accession,DNA Accession,AMR Gene Family,Drug Class,Resistance Mechanism,CARD Short Name
0,ARO:3005099,43314,6143,3831,23S rRNA (adenine(2058)-N(6))-methyltransferas...,23S rRNA (adenine(2058)-N(6))-methyltransferas...,AAB60941.1,AF002716.1,Erm 23S ribosomal RNA methyltransferase,lincosamide antibiotic;macrolide antibiotic;st...,antibiotic target alteration,Spyo_ErmA_MLSb
1,ARO:3002523,38923,8144,1781,AAC(2')-Ia,AAC(2')-Ia,AAA03550.1,L06156.2,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ia
2,ARO:3002524,38924,85,746,AAC(2')-Ib,AAC(2')-Ib,AAC44793.1,U41471.1,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ib
3,ARO:3002525,38925,4719,1246,AAC(2')-Ic,AAC(2')-Ic,CCP42991.1,AL123456.3,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ic
4,ARO:3002526,38926,228,1415,AAC(2')-Id,AAC(2')-Id,AAB41701.1,U72743.1,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Id
...,...,...,...,...,...,...,...,...,...,...,...,...
6440,ARO:3003064,39498,536,350,ykkD,ykkD,CAB13167.1,AL009126.1,small multidrug resistance (SMR) antibiotic ef...,aminoglycoside antibiotic;phenicol antibiotic;...,antibiotic efflux,ykkD
6441,ARO:3009096,47888,11341,8514,YOC-1,YOC-1,QIU91361.1,CP050811.1,YOC beta-lactamase,cephalosporin,antibiotic inactivation,YOC-1
6442,ARO:3003952,40722,5261,2424,YojI,YojI,AAC75271.1,U00096.3,ATP-binding cassette (ABC) antibiotic efflux pump,peptide antibiotic,antibiotic efflux,YojI
6443,ARO:3005035,43233,6072,3779,YRC-1,YRC-1,ABA70720.1,DQ185144.1,YRC beta-lactamase,cephalosporin;penicillin beta-lactam,antibiotic inactivation,YRC-1


In [19]:
categories

,ARO Category,ARO Accession,ARO Name
0,AMR Gene Family,ARO:3004272,16S rRNA methyltransferase (A1408)
1,AMR Gene Family,ARO:3004271,16S rRNA methyltransferase (G1405)
2,AMR Gene Family,ARO:3003666,16s rRNA with mutation conferring resistance t...
3,AMR Gene Family,ARO:3003976,16S rRNA with mutation conferring resistance t...
4,AMR Gene Family,ARO:3003667,16s rRNA with mutation conferring resistance t...
...,...,...,...
647,Resistance Mechanism,ARO:0001002,antibiotic target replacement
648,Resistance Mechanism,ARO:3007115,modification to cell morphology
649,Resistance Mechanism,ARO:3000244,reduced permeability to antibiotic
650,Resistance Mechanism,ARO:3003764,resistance by absence


In [20]:
# practicing merge () on dummy files first

import pandas as pd

students = pd.DataFrame({
    "Student": ["Ali", "Sara", "John", "Ahmed", "Fatima"],
    "Department": ["Biology", "Biology", "Chemistry", "Physics", "Chemistry"],
    "Marks": [88, 92, 76, 85, 90]
})

students

,Student,Department,Marks
0,Ali,Biology,88
1,Sara,Biology,92
2,John,Chemistry,76
3,Ahmed,Physics,85
4,Fatima,Chemistry,90


In [25]:
students.groupby("Department").size()


Department
Biology      2
Chemistry    2
Physics      1
dtype: int64

In [28]:
students.groupby("Department")["Marks"].mean()

Department
Biology      90.0
Chemistry    83.0
Physics      85.0
Name: Marks, dtype: float64

In [29]:
students.groupby("Department")["Marks"].max()

Department
Biology      92
Chemistry    90
Physics      85
Name: Marks, dtype: int64

In [30]:
students.groupby("Department")["Student"].count()

Department
Biology      2
Chemistry    2
Physics      1
Name: Student, dtype: int64

# merge ()  
- merge() is equivalent to Excel's VLOOKUP or XLOOKUP: it matches rows based on a shared key and brings over related information.

In [31]:
advisors = pd.DataFrame({
    "Department": ["Biology", "Chemistry", "Physics"],
    "Advisor": ["Dr. Khan", "Dr. Smith", "Dr. Lee"]
})

advisors

,Department,Advisor
0,Biology,Dr. Khan
1,Chemistry,Dr. Smith
2,Physics,Dr. Lee


In [32]:
students.merge(advisors, on= "Department")

,Student,Department,Marks,Advisor
0,Ali,Biology,88,Dr. Khan
1,Sara,Biology,92,Dr. Khan
2,John,Chemistry,76,Dr. Smith
3,Ahmed,Physics,85,Dr. Lee
4,Fatima,Chemistry,90,Dr. Smith


## The difference in one sentence

If you remember only one thing today, make it this:

groupby() = "Take one table and summarize it."
merge() = "Take two tables and combine them using a shared key."

# Applying groubly() and Merge () on CARD data. 

In [33]:
#Question 1

#How many resistance genes are in each drug class?

In [35]:
aro.groupby("Drug Class").size().sort_values(ascending = False)

Drug Class
cephalosporin                                                                                                                                                                                                                 943
carbapenem;penicillin beta-lactam                                                                                                                                                                                             821
carbapenem;cephalosporin;monobactam                                                                                                                                                                                           638
penicillin beta-lactam                                                                                                                                                                                                        583
carbapenem;cephalosporin;penicillin beta-lactam                                      

In [36]:
#Question 2

#How many genes belong to each AMR Gene Family?

aro.groupby("AMR Gene Family").size().sort_values(ascending = False)

AMR Gene Family
PDC beta-lactamase                                                       639
OXA beta-lactamase;OXA-51-like beta-lactamase                            383
ADC beta-lactamases pending classification for carbapenemase activity    274
CTX-M beta-lactamase                                                     267
KPC beta-lactamase                                                       229
                                                                        ... 
VHW beta-lactamase                                                         1
VHH beta-lactamase                                                         1
EAM beta-lactamase                                                         1
VCC beta-lactamase                                                         1
viomycin phosphotransferase                                                1
Length: 540, dtype: int64

In [37]:
#Question 3

#How many resistance mechanisms are represented?

aro.groupby("Resistance Mechanism").size()

Resistance Mechanism
antibiotic efflux                                                                     296
antibiotic efflux;antibiotic target alteration                                         28
antibiotic efflux;antibiotic target alteration;reduced permeability to antibiotic       2
antibiotic efflux;antibiotic target alteration;resistance by absence                    4
antibiotic efflux;reduced permeability to antibiotic                                    4
antibiotic inactivation                                                              5245
antibiotic inactivation;antibiotic target alteration                                    2
antibiotic target alteration                                                          580
antibiotic target alteration;antibiotic target replacement                             10
antibiotic target protection                                                          176
antibiotic target replacement                                                  

In [38]:
# looking for missing values 

aro.isna().sum()

ARO Accession            0
CVTERM ID                0
Model Sequence ID        0
Model ID                 0
Model Name               0
ARO Name                 0
Protein Accession       87
DNA Accession            0
AMR Gene Family          0
Drug Class               0
Resistance Mechanism     0
CARD Short Name          0
dtype: int64

In [42]:
aro[aro["Drug Class"] == "fluoroquinolone antibiotic"]

,ARO Accession,CVTERM ID,Model Sequence ID,Model ID,Model Name,ARO Name,Protein Accession,DNA Accession,AMR Gene Family,Drug Class,Resistance Mechanism,CARD Short Name
144,ARO:3004574,42490,5314,3278,Acinetobacter baumannii AbaQ,Acinetobacter baumannii AbaQ,ACJ41547.2,CP001182.2,major facilitator superfamily (MFS) antibiotic...,fluoroquinolone antibiotic,antibiotic efflux,Abau_AbaQ
150,ARO:3003818,40508,6039,2315,Acinetobacter baumannii parC conferring resist...,Acinetobacter baumannii parC conferring resist...,WP_000202252.1,NZ_CP009257.1,fluoroquinolone resistant parC,fluoroquinolone antibiotic,antibiotic target alteration,Abau_parC_FLO
1697,ARO:3000027,36036,411,1757,emrA,emrA,BAA16547.1,AP009048.1,major facilitator superfamily (MFS) antibiotic...,fluoroquinolone antibiotic,antibiotic efflux,emrA
1698,ARO:3000074,36213,20,1847,emrB,emrB,AAC75733.1,U00096.1,major facilitator superfamily (MFS) antibiotic...,fluoroquinolone antibiotic,antibiotic efflux,emrB
1701,ARO:3000516,36655,5286,1330,emrR,emrR,AAC75731.1,U00096.3,major facilitator superfamily (MFS) antibiotic...,fluoroquinolone antibiotic,antibiotic efflux,emrR
...,...,...,...,...,...,...,...,...,...,...,...,...
5836,ARO:3004667,42714,5552,3367,Staphylococcus aureus norA,Staphylococcus aureus norA,BAA14147.1,D90119.1,major facilitator superfamily (MFS) antibiotic...,fluoroquinolone antibiotic,antibiotic efflux,Saur_norA
5837,ARO:3003312,39896,6038,212,Staphylococcus aureus parC conferring resistan...,Staphylococcus aureus parC conferring resistan...,WP_001289569.1,NC_002951.2,fluoroquinolone resistant parC,fluoroquinolone antibiotic,antibiotic target alteration,Saur_parC_FLO
5839,ARO:3003315,39899,5400,335,Staphylococcus aureus parE conferring resistan...,Staphylococcus aureus parE conferring resistan...,CAG40364.1,BX571856.1,fluoroquinolone resistant parE,fluoroquinolone antibiotic,antibiotic target alteration,Saur_parE_FLO
5851,ARO:3003311,39895,6037,660,Streptococcus pneumoniae parC conferring resis...,Streptococcus pneumoniae parC conferring resis...,NP_358351.1,NC_003098.1,fluoroquinolone resistant parC,fluoroquinolone antibiotic,antibiotic target alteration,Spne_parC_FLO


In [43]:
counts 

Drug Class
cephalosporin                                                                                                                                                                                                                 943
carbapenem;penicillin beta-lactam                                                                                                                                                                                             821
carbapenem;cephalosporin;monobactam                                                                                                                                                                                           638
penicillin beta-lactam                                                                                                                                                                                                        583
carbapenem;cephalosporin;penicillin beta-lactam                                      

In [44]:
counts = aro.groupby("Drug Class").size().sort_values(ascending=False)

counts.head(15)

Drug Class
cephalosporin                                                 943
carbapenem;penicillin beta-lactam                             821
carbapenem;cephalosporin;monobactam                           638
penicillin beta-lactam                                        583
carbapenem;cephalosporin;penicillin beta-lactam               579
cephalosporin;penicillin beta-lactam                          461
cephalosporin;monobactam;penicillin beta-lactam               312
carbapenem;cephalosporin;monobactam;penicillin beta-lactam    287
aminoglycoside antibiotic                                     200
carbapenem                                                    184
peptide antibiotic                                            168
fluoroquinolone antibiotic                                    147
glycopeptide antibiotic                                        86
tetracycline antibiotic                                        82
cephalosporin;monobactam                                       74

### What does 943 mean?

It does not mean there are 943 different cephalosporin antibiotics.

It means:

There are 943 rows in the CARD database whose Drug Class field is exactly "cephalosporin".

Remember, groupby().size() counts rows, not unique drugs or genes.

In [45]:
aro[aro["Drug Class"] == "cephalosporin"]

,ARO Accession,CVTERM ID,Model Sequence ID,Model ID,Model Name,ARO Name,Protein Accession,DNA Accession,AMR Gene Family,Drug Class,Resistance Mechanism,CARD Short Name
328,ARO:3003847,40544,5287,2334,ADC-1,ADC-1,CAB77444.1,AJ009979.1,ADC beta-lactamase without carbapenemase activity,cephalosporin,antibiotic inactivation,ADC-1
329,ARO:3003855,40554,5246,3314,ADC-10,ADC-10,ABI18382.1,DQ883640.1,ADC beta-lactamase without carbapenemase activity,cephalosporin,antibiotic inactivation,ADC-10
330,ARO:3006280,44742,6551,4176,ADC-100,ADC-100,WP_068981607.1,NG_051439.1,ADC beta-lactamases pending classification for...,cephalosporin,antibiotic inactivation,ADC-100
331,ARO:3006281,44743,6552,4177,ADC-101,ADC-101,WP_068981608.1,NG_051440.1,ADC beta-lactamases pending classification for...,cephalosporin,antibiotic inactivation,ADC-101
332,ARO:3006282,44744,6553,4178,ADC-102,ADC-102,WP_068981609.1,NG_051441.1,ADC beta-lactamases pending classification for...,cephalosporin,antibiotic inactivation,ADC-102
...,...,...,...,...,...,...,...,...,...,...,...,...
5812,ARO:3002494,38894,8219,1878,SRT-2,SRT-2,AAS07017.3,AY524276.3,SRT beta-lactamase,cephalosporin,antibiotic inactivation,SRT-2
5813,ARO:3006853,45315,8039,5664,SRT-3,SRT-3,WP_033642964.1,NG_070754.1,SRT beta-lactamase,cephalosporin,antibiotic inactivation,SRT-3
5814,ARO:3009065,47857,11310,8483,SRT-4,SRT-4,UBX54502.1,OK340848.1,SRT beta-lactamase,cephalosporin,antibiotic inactivation,SRT-4
5816,ARO:3006999,45461,8040,5665,SST-1,SST-1,WP_063864750.1,NG_050144.1,SST beta-lactamase,cephalosporin,antibiotic inactivation,SST-1


In [46]:
aro[aro["Drug Class"] == "cephalosporin"].shape

(943, 12)

In [47]:
aro[aro["Drug Class"] == "cephalosporin"].head()

,ARO Accession,CVTERM ID,Model Sequence ID,Model ID,Model Name,ARO Name,Protein Accession,DNA Accession,AMR Gene Family,Drug Class,Resistance Mechanism,CARD Short Name
328,ARO:3003847,40544,5287,2334,ADC-1,ADC-1,CAB77444.1,AJ009979.1,ADC beta-lactamase without carbapenemase activity,cephalosporin,antibiotic inactivation,ADC-1
329,ARO:3003855,40554,5246,3314,ADC-10,ADC-10,ABI18382.1,DQ883640.1,ADC beta-lactamase without carbapenemase activity,cephalosporin,antibiotic inactivation,ADC-10
330,ARO:3006280,44742,6551,4176,ADC-100,ADC-100,WP_068981607.1,NG_051439.1,ADC beta-lactamases pending classification for...,cephalosporin,antibiotic inactivation,ADC-100
331,ARO:3006281,44743,6552,4177,ADC-101,ADC-101,WP_068981608.1,NG_051440.1,ADC beta-lactamases pending classification for...,cephalosporin,antibiotic inactivation,ADC-101
332,ARO:3006282,44744,6553,4178,ADC-102,ADC-102,WP_068981609.1,NG_051441.1,ADC beta-lactamases pending classification for...,cephalosporin,antibiotic inactivation,ADC-102


# Merge ()

In [48]:
categories = pd.read_csv("data/card/aro_categories.tsv", sep="\t")

In [49]:
aro.columns.tolist()

['ARO Accession',
 'CVTERM ID',
 'Model Sequence ID',
 'Model ID',
 'Model Name',
 'ARO Name',
 'Protein Accession',
 'DNA Accession',
 'AMR Gene Family',
 'Drug Class',
 'Resistance Mechanism',
 'CARD Short Name']

In [50]:
categories.columns.tolist()

['ARO Category', 'ARO Accession', 'ARO Name']

In [63]:
categories.head()

,ARO Category,ARO Accession,ARO Name
0,AMR Gene Family,ARO:3004272,16S rRNA methyltransferase (A1408)
1,AMR Gene Family,ARO:3004271,16S rRNA methyltransferase (G1405)
2,AMR Gene Family,ARO:3003666,16s rRNA with mutation conferring resistance t...
3,AMR Gene Family,ARO:3003976,16S rRNA with mutation conferring resistance t...
4,AMR Gene Family,ARO:3003667,16s rRNA with mutation conferring resistance t...


In [64]:
categories.isna().sum()

ARO Category     0
ARO Accession    0
ARO Name         0
dtype: int64

In [60]:
merged = aro.merge(categories, on="ARO Accession", how="left")

In [52]:
merged = aro.merge(categories, on="ARO Accession", how="left")

In [61]:
merged.shape

(6445, 14)

In [62]:
merged.head()

,ARO Accession,CVTERM ID,Model Sequence ID,Model ID,Model Name,ARO Name_x,Protein Accession,DNA Accession,AMR Gene Family,Drug Class,Resistance Mechanism,CARD Short Name,ARO Category,ARO Name_y
0,ARO:3005099,43314,6143,3831,23S rRNA (adenine(2058)-N(6))-methyltransferas...,23S rRNA (adenine(2058)-N(6))-methyltransferas...,AAB60941.1,AF002716.1,Erm 23S ribosomal RNA methyltransferase,lincosamide antibiotic;macrolide antibiotic;st...,antibiotic target alteration,Spyo_ErmA_MLSb,NaN,NaN
1,ARO:3002523,38923,8144,1781,AAC(2')-Ia,AAC(2')-Ia,AAA03550.1,L06156.2,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ia,NaN,NaN
2,ARO:3002524,38924,85,746,AAC(2')-Ib,AAC(2')-Ib,AAC44793.1,U41471.1,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ib,NaN,NaN
3,ARO:3002525,38925,4719,1246,AAC(2')-Ic,AAC(2')-Ic,CCP42991.1,AL123456.3,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ic,NaN,NaN
4,ARO:3002526,38926,228,1415,AAC(2')-Id,AAC(2')-Id,AAB41701.1,U72743.1,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Id,NaN,NaN


In [65]:
merged.isna().sum()

ARO Accession              0
CVTERM ID                  0
Model Sequence ID          0
Model ID                   0
Model Name                 0
ARO Name_x                 0
Protein Accession         87
DNA Accession              0
AMR Gene Family            0
Drug Class                 0
Resistance Mechanism       0
CARD Short Name            0
ARO Category            6444
ARO Name_y              6444
dtype: int64

In [66]:
merged.shape

(6445, 14)

In [67]:
aro.shape

(6445, 12)

In [68]:
aro["ARO Accession"].dtype

<StringDtype(storage='python', na_value=nan)>

In [69]:
aro["ARO Accession"].nunique()

6440

In [70]:
categories["ARO Accession"].nunique()

652

In [71]:
set(aro["ARO Accession"]) & set(categories["ARO Accession"])

{'ARO:3007057'}

In [72]:
len(set(aro["ARO Accession"]) & set(categories["ARO Accession"]))len(set(aro["ARO Accession"]) & set(categories["ARO Accession"]))

1

In [73]:
categories.head(10)


,ARO Category,ARO Accession,ARO Name
0,AMR Gene Family,ARO:3004272,16S rRNA methyltransferase (A1408)
1,AMR Gene Family,ARO:3004271,16S rRNA methyltransferase (G1405)
2,AMR Gene Family,ARO:3003666,16s rRNA with mutation conferring resistance t...
3,AMR Gene Family,ARO:3003976,16S rRNA with mutation conferring resistance t...
4,AMR Gene Family,ARO:3003667,16s rRNA with mutation conferring resistance t...
5,AMR Gene Family,ARO:3003668,16s rRNA with mutation conferring resistance t...
6,AMR Gene Family,ARO:3003669,16S rRNA with mutation conferring resistance t...
7,AMR Gene Family,ARO:3004936,23s rRNA with mutation conferring resistance t...
8,AMR Gene Family,ARO:3004187,23S rRNA with mutation conferring resistance t...
9,AMR Gene Family,ARO:3004057,23S rRNA with mutation conferring resistance t...


In [74]:
categories.shape

(652, 3)

In [75]:
categories["ARO Accession"] = categories["ARO Accession"].str.replace("ARO:", "")

In [76]:
import pandas as pd

df = pd.read_csv("data/card/aro_index.tsv", sep="\t")

# groupby: how many entries per drug class?
counts = df.groupby("Drug Class").size().sort_values(ascending=False)
print(counts.head(15))

Drug Class
cephalosporin                                                 943
carbapenem;penicillin beta-lactam                             821
carbapenem;cephalosporin;monobactam                           638
penicillin beta-lactam                                        583
carbapenem;cephalosporin;penicillin beta-lactam               579
cephalosporin;penicillin beta-lactam                          461
cephalosporin;monobactam;penicillin beta-lactam               312
carbapenem;cephalosporin;monobactam;penicillin beta-lactam    287
aminoglycoside antibiotic                                     200
carbapenem                                                    184
peptide antibiotic                                            168
fluoroquinolone antibiotic                                    147
glycopeptide antibiotic                                        86
tetracycline antibiotic                                        82
cephalosporin;monobactam                                       74

In [78]:
merged = df.merge(categories, on="ARO Accession", how="left")  # adjust column name to match what you actually see
print(merged.shape)


(6445, 14)


In [79]:
merged


,ARO Accession,CVTERM ID,Model Sequence ID,Model ID,Model Name,ARO Name_x,Protein Accession,DNA Accession,AMR Gene Family,Drug Class,Resistance Mechanism,CARD Short Name,ARO Category,ARO Name_y
0,ARO:3005099,43314,6143,3831,23S rRNA (adenine(2058)-N(6))-methyltransferas...,23S rRNA (adenine(2058)-N(6))-methyltransferas...,AAB60941.1,AF002716.1,Erm 23S ribosomal RNA methyltransferase,lincosamide antibiotic;macrolide antibiotic;st...,antibiotic target alteration,Spyo_ErmA_MLSb,NaN,NaN
1,ARO:3002523,38923,8144,1781,AAC(2')-Ia,AAC(2')-Ia,AAA03550.1,L06156.2,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ia,NaN,NaN
2,ARO:3002524,38924,85,746,AAC(2')-Ib,AAC(2')-Ib,AAC44793.1,U41471.1,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ib,NaN,NaN
3,ARO:3002525,38925,4719,1246,AAC(2')-Ic,AAC(2')-Ic,CCP42991.1,AL123456.3,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ic,NaN,NaN
4,ARO:3002526,38926,228,1415,AAC(2')-Id,AAC(2')-Id,AAB41701.1,U72743.1,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Id,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6440,ARO:3003064,39498,536,350,ykkD,ykkD,CAB13167.1,AL009126.1,small multidrug resistance (SMR) antibiotic ef...,aminoglycoside antibiotic;phenicol antibiotic;...,antibiotic efflux,ykkD,NaN,NaN
6441,ARO:3009096,47888,11341,8514,YOC-1,YOC-1,QIU91361.1,CP050811.1,YOC beta-lactamase,cephalosporin,antibiotic inactivation,YOC-1,NaN,NaN
6442,ARO:3003952,40722,5261,2424,YojI,YojI,AAC75271.1,U00096.3,ATP-binding cassette (ABC) antibiotic efflux pump,peptide antibiotic,antibiotic efflux,YojI,NaN,NaN
6443,ARO:3005035,43233,6072,3779,YRC-1,YRC-1,ABA70720.1,DQ185144.1,YRC beta-lactamase,cephalosporin;penicillin beta-lactam,antibiotic inactivation,YRC-1,NaN,NaN


In [80]:
categories

,ARO Category,ARO Accession,ARO Name
0,AMR Gene Family,3004272,16S rRNA methyltransferase (A1408)
1,AMR Gene Family,3004271,16S rRNA methyltransferase (G1405)
2,AMR Gene Family,3003666,16s rRNA with mutation conferring resistance t...
3,AMR Gene Family,3003976,16S rRNA with mutation conferring resistance t...
4,AMR Gene Family,3003667,16s rRNA with mutation conferring resistance t...
...,...,...,...
647,Resistance Mechanism,0001002,antibiotic target replacement
648,Resistance Mechanism,3007115,modification to cell morphology
649,Resistance Mechanism,3000244,reduced permeability to antibiotic
650,Resistance Mechanism,3003764,resistance by absence


In [82]:
categories["ARO Accession"] = categories["ARO Accession"].str.replace("ARO:", "")

In [83]:
categories

,ARO Category,ARO Accession,ARO Name
0,AMR Gene Family,3004272,16S rRNA methyltransferase (A1408)
1,AMR Gene Family,3004271,16S rRNA methyltransferase (G1405)
2,AMR Gene Family,3003666,16s rRNA with mutation conferring resistance t...
3,AMR Gene Family,3003976,16S rRNA with mutation conferring resistance t...
4,AMR Gene Family,3003667,16s rRNA with mutation conferring resistance t...
...,...,...,...
647,Resistance Mechanism,0001002,antibiotic target replacement
648,Resistance Mechanism,3007115,modification to cell morphology
649,Resistance Mechanism,3000244,reduced permeability to antibiotic
650,Resistance Mechanism,3003764,resistance by absence


In [84]:
merged = df.merge(categories, on="ARO Accession", how="left")  # adjust column name to match what you actually see
print(merged.shape)

(6445, 14)


In [91]:
merged = merged.drop(columns='ARO Category')

In [92]:
merged = merged.drop(columns='ARO Name_y')

In [93]:
merged

,ARO Accession,CVTERM ID,Model Sequence ID,Model ID,Model Name,ARO Name_x,Protein Accession,DNA Accession,AMR Gene Family,Drug Class,Resistance Mechanism,CARD Short Name
0,ARO:3005099,43314,6143,3831,23S rRNA (adenine(2058)-N(6))-methyltransferas...,23S rRNA (adenine(2058)-N(6))-methyltransferas...,AAB60941.1,AF002716.1,Erm 23S ribosomal RNA methyltransferase,lincosamide antibiotic;macrolide antibiotic;st...,antibiotic target alteration,Spyo_ErmA_MLSb
1,ARO:3002523,38923,8144,1781,AAC(2')-Ia,AAC(2')-Ia,AAA03550.1,L06156.2,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ia
2,ARO:3002524,38924,85,746,AAC(2')-Ib,AAC(2')-Ib,AAC44793.1,U41471.1,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ib
3,ARO:3002525,38925,4719,1246,AAC(2')-Ic,AAC(2')-Ic,CCP42991.1,AL123456.3,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Ic
4,ARO:3002526,38926,228,1415,AAC(2')-Id,AAC(2')-Id,AAB41701.1,U72743.1,AAC(2'),aminoglycoside antibiotic,antibiotic inactivation,AAC(2')-Id
...,...,...,...,...,...,...,...,...,...,...,...,...
6440,ARO:3003064,39498,536,350,ykkD,ykkD,CAB13167.1,AL009126.1,small multidrug resistance (SMR) antibiotic ef...,aminoglycoside antibiotic;phenicol antibiotic;...,antibiotic efflux,ykkD
6441,ARO:3009096,47888,11341,8514,YOC-1,YOC-1,QIU91361.1,CP050811.1,YOC beta-lactamase,cephalosporin,antibiotic inactivation,YOC-1
6442,ARO:3003952,40722,5261,2424,YojI,YojI,AAC75271.1,U00096.3,ATP-binding cassette (ABC) antibiotic efflux pump,peptide antibiotic,antibiotic efflux,YojI
6443,ARO:3005035,43233,6072,3779,YRC-1,YRC-1,ABA70720.1,DQ185144.1,YRC beta-lactamase,cephalosporin;penicillin beta-lactam,antibiotic inactivation,YRC-1


In [94]:
merged = df.merge(categories, on="ARO Accession", how="left")  # adjust column name to match what you actually see
print(merged.shape)

(6445, 14)


In [96]:
categories

,ARO Category,ARO Accession,ARO Name
0,AMR Gene Family,3004272,16S rRNA methyltransferase (A1408)
1,AMR Gene Family,3004271,16S rRNA methyltransferase (G1405)
2,AMR Gene Family,3003666,16s rRNA with mutation conferring resistance t...
3,AMR Gene Family,3003976,16S rRNA with mutation conferring resistance t...
4,AMR Gene Family,3003667,16s rRNA with mutation conferring resistance t...
...,...,...,...
647,Resistance Mechanism,0001002,antibiotic target replacement
648,Resistance Mechanism,3007115,modification to cell morphology
649,Resistance Mechanism,3000244,reduced permeability to antibiotic
650,Resistance Mechanism,3003764,resistance by absence


In [98]:
len(set(df["ARO Accession"]) & set(categories["ARO Accession"]))

0

In [100]:
set(df["ARO Accession"]) & set(categories["ARO Accession"])

set()

In [101]:
genes = pd.DataFrame({"gene": ["acrR", "parE", "ompR", "yjgA"], "mechanism": ["efflux regulation", "target alteration", "regulatory", "unknown"]})
lengths = pd.DataFrame({"gene": ["acrR", "parE", "ompR", "yjgA"], "length_bp": [624, 2121, 687, 351]})
genes.merge(lengths, on="gene")


,gene,mechanism,length_bp
0,acrR,efflux regulation,624
1,parE,target alteration,2121
2,ompR,regulatory,687
3,yjgA,unknown,351
